design/baremetal-operator/firmware-interface.md

```diff
@@ -0,0 +1,225 @@
+
+When provisioning new machines to clusters, it may be necessary to upgrade
+or downgrade the servers BIOS and firmware to specific versions to ensure
+the server is configured as it is in their validated pattern.
+
+Redfish has a common API that allow users to execute a firmware update, the
+[proposal](https://specs.openstack.org/openstack/ironic-specs/priorities/2023-2-workitems.html#firmware-updates)
+was accepted in the Ironic community, and is already implemented.
+
+### Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow firmware updates for the BMC and BIOS in hardware only compatible
+  with Redfish.
+
+### Non-Goals
+
+- Update firmware using non Redfish means
+- Update firmware that are not for BMC and BIOS
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator I want to install specific versions of firmware in my
+machines before installing the Operating System.
+
+## Design Details
+
+This proposes a new Custom Resource Definition (CRD) to store the information
+about the firmware components from Ironic. The initial version stores only
+the firmware information about BMC and BIOS, we may expand this to other
+components in the future.
+The new CRD is named `HostFirmwareComponents` and will consist of the
+following:
+
+- `components` - the current firmware components and their information,
+  retrieved from Ironic via the [Firmware API](https://docs.openstack.org/api-ref/baremetal/#node-firmware-nodes)
+  will be stored in the `Status` section.
+- `updates` - firmware components to be updated via Ironic will be
+  stored in the `Spec` section. It will be empty when the CRD is created.
+
+The firmware components are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the components
+are updated. The BMO manages the data as follows:
+
+- The node first transitions to manageable during the `Registering` state, so
+  at the end of that state `components` will be populated.
+- Firmware components can be updated during the `Preparing` state, so at the
+  end of that state the components will also be retrieved and used to update
+  `components`.
+
+A user can  update `updates` to specify the desired firmware of each
+component. The BMO will detect changes on it by comparing the name/url pairs
+to the values in `Status`. When a change is detected, the BMO will add the
+new values to the Ironic clean-steps API in the Preparing state, when building
+the manual clean steps, the Host will re-enter this state from Ready/Available
+state whenever its config differs from the last stored one.
+
+After executing the cleaning, Ironic will re-read the information about the
+firmware components and cache them, the new information can be retrieved by
+the BMO and used to update `components`.
+
+If the update fails we won’t keep trying to reconcile, the BMO will put the
+node in a `Error` state. We will allow deletion of the BMH when in this state.
```

Comment:
New steps are cheaper in BMO than in Ironic, so maybe we get specific and actually call the state `FirmwareUpgradeError`. The allowed transitions will be to `Preparing` and `Deleting`. I suggest, we won't do anything else to the host in case of deleting (in case we've just bricked the machine, lol).

I think we also need some sort of a normal transition. Something like "I have checked that the Machine is still operational, but I don't want to do any more updates". Probably by deleting "updates" from the CR? What do you think?

Reply:
I've updated and added more information regarding this, tks!